src/solver.rs

```diff
@@ -338,14 +352,76 @@ impl<P: Package, VS: VersionSet> OfflineDependencyProvider<P, VS> {
         version: impl Into<VS::V>,
         dependencies: I,
     ) {
-        let package_deps = dependencies.into_iter().collect();
         let v = version.into();
-        *self
+        let entries = self
+            .dependencies
+            .entry(package)
+            .or_default()
+            .entry(v)
+            .or_default();
+        for (dep, range) in dependencies.into_iter() {
+            entries.insert(dep, Requirement::Required(range));
+        }
+    }
+
+    /// Registers the dependencies of a package and version pair that may *not* be included in the final
+    /// solution but if it is included in the solution (through another package for example) it must meet
+    /// the requirements defined in this function call..
+    /// Dependencies must be added with a single call to
+    /// [add_dependencies](OfflineDependencyProvider::add_dependencies),
+    /// [add_constraints](OfflineDependencyProvider::add_constraints), or
+    /// [add_requirements](OfflineDependencyProvider::add_requirements).
+    /// All subsequent calls to any of those functions for a given package version pair will replace the
+    /// dependencies by the new ones.
+    ///
+    /// The API does not allow to add dependencies one at a time to uphold an assumption that
+    /// [OfflineDependencyProvider.get_dependencies(p, v)](OfflineDependencyProvider::get_dependencies)
+    /// provides all dependencies of a given package (p) and version (v) pair.
+    pub fn add_constraints<I: IntoIterator<Item = (P, VS)>>(
+        &mut self,
+        package: P,
+        version: impl Into<VS::V>,
+        dependencies: I,
+    ) {
+        let v = version.into();
+        let entries = self
+            .dependencies
+            .entry(package)
+            .or_default()
+            .entry(v)
+            .or_default();
+        for (package, constraint) in dependencies.into_iter() {
+            entries.insert(package, Requirement::Constrained(constraint));
+        }
+    }
+
+    /// Registers the dependencies of a package and version pair.
+    /// Dependencies must be added with a single call to
+    /// [add_dependencies](OfflineDependencyProvider::add_dependencies),
+    /// [add_constraints](OfflineDependencyProvider::add_constraints), or
+    /// [add_requirements](OfflineDependencyProvider::add_requirements).
+    /// All subsequent calls to any of those functions for a given package version pair will replace the
```

Comment:
This is no longer true. At a minimum these comments need to be updated. These calls now append items to an existing list.
@mpizenberg, do you have a strong preference for the all at once API?

Reply:
Indeed, I'm not a big fan of making it easy for people to make mistakes. I'd say we probably make `add_dependency` and `add_constraint` private or remove them alltogether, and only keep the `add_requirements` function. And the new `add_requirements` function should use `=` instead of `insert`. It doesn't make sense to have both a required dep and a constraint for the same dependency package.